File: aoc2017/day1.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <stdint.h>

char *loaddata(const char *filename, int *nread) {
  FILE *f = fopen(filename, "r");
  char *out = NULL;

  int cread = 0;
  char c = 0;
  while (fread(&c, sizeof (char), 1, f) != 0) {
    if (c != '\n') {
      cread++;
      out = realloc(out, cread * sizeof (char));
      out[cread-1] = c;
    }
  }

  *nread = cread;
  fclose(f);
  return out;
}

void pt1(char *data, int nread) {
  int sum = 0;
  for (int i = 0; i < nread; i++) {
    int n = data[i];
    int m = data[(i+1) % nread];
    if (n == m) {
      sum += n - 48;
    }
  }

  printf("pt1: %d\n", sum);
}

void pt2(char *data, int nread) {
  int sum = 0;
  for (int i = 0; i < nread; i++) {
    int n = data[i];
    int m = data[(i + nread/2) % nread];
    if (n == m) {
      sum += n - 48;
    }
  }
  printf("pt2: %d\n", sum);
}

uint64_t us() {
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return tv.tv_sec * 1000000 + tv.tv_usec;
}

int main(int argc, char *argv[]) {
  clock_t start_c = clock();
  uint64_t start = us();

  int nread = 0;
  char *data = loaddata("day1.txt", &nread);

  printf("us = %llu\n", us());
  pt1(data, nread);
  pt2(data, nread);


  free(data);

  clock_t stop_c = clock();
  uint64_t stop = us();

  printf("took %llu us (via timeval)\n", stop-start);
  printf("took %0.0llf us (via clock_t)\n", ((double) stop_c-(double)start_c)/CLOCKS_PER_SEC * 1000000.0);

  return 0;
}

File: aoc2017/day1.py
import timeit

from si_prefix import si_format as si


def pt1(data):
  dlen = len(data)
  return sum([n for i, n in enumerate(data) if n == data[(i+1) % dlen]])

def pt2(data):
  dlen = len(data)
  step = dlen//2
  return sum([n for i, n in enumerate(data) if n == data[(i+step) % dlen]])

if __name__ == '__main__':
  with open('day1.txt', 'r') as f:
    data = [int(n) for n in f.read().strip()]


  pt1_time = timeit.timeit(lambda: pt1(data), number=1000)
  pt2_time = timeit.timeit(lambda: pt2(data), number=1000)
  print(f'pt1: {pt1(data):10} | {si(pt1_time)}s')
  print(f'pt2: {pt2(data):10} | {si(pt2_time)}s')
